File: pippip-native-ios/More/ContactPolicyCell.m
//
//  ContactPolicyCell.m
//  pippip-native-ios
//
//  Created by Steve Brenneis on 2/23/18.
//  Copyright © 2018 seComm. All rights reserved.
//

#import "ContactPolicyCell.h"
#import "pippip_native_ios-Swift.h"
#import "Configurator.h"
#import "ContactManager.h"
#import "Notifications.h"
#import "ApplicationSingleton.h"

@interface ContactPolicyCell ()
{
    NSString *selectedPolicy;
    NSString *currentPolicy;
    Configurator *config;
    ContactManager *contactManager;
}

@property (weak, nonatomic) IBOutlet UISwitch *contactPolicySwitch;

@end

@implementation ContactPolicyCell

+ (MoreCellItem*)cellItem {

    MoreCellItem *item = [[MoreCellItem alloc] init];
    item.cellReuseId = @"ContactPolicyCell";
    item.cellHeight = 65.0;
    return item;

}

- (void)awakeFromNib {
    [super awakeFromNib];
    // Initialization code

    contactManager = [[ContactManager alloc] init];
    config = [ApplicationSingleton instance].config;
    currentPolicy = [config getContactPolicy];
    if ([currentPolicy isEqualToString:@"public"]) {
        [_contactPolicySwitch setOn:YES animated:YES];
    }
    else {
        [_contactPolicySwitch setOn:NO animated:YES];
    }

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(policyUpdated:)
                                                 name:POLICY_UPDATED
                                               object:nil];

}

- (void)setSelected:(BOOL)selected animated:(BOOL)animated {
    [super setSelected:selected animated:animated];

    // Configure the view for the selected state
}

- (void)policyUpdated:(NSNotification*)notification {

    NSDictionary *info = notification.userInfo;
    if (info != nil) {
        NSString *result = info[@"result"];
        if ([result isEqualToString:@"policySet"]) {
            currentPolicy = selectedPolicy;
            [config setContactPolicy:selectedPolicy];
            NSLog(@"Contact policy set to %@", currentPolicy);
        }
        dispatch_async(dispatch_get_main_queue(), ^{
            if ([currentPolicy isEqualToString:@"public"]) {
                [_contactPolicySwitch setOn:YES animated:YES];
            }
            else {
                [_contactPolicySwitch setOn:NO animated:YES];
            }
        });
        NSMutableDictionary *info = [NSMutableDictionary dictionary];
        info[@"policy"] = currentPolicy;
        [AsyncNotifier notifyWithName:POLICY_CHANGED object:nil userInfo:info];
    }

}

- (IBAction)policyChanged:(UISwitch *)sender {

    if (sender.on) {
        selectedPolicy = @"public";
    }
    else {
        selectedPolicy = @"whitelist";
    }
    [contactManager setContactPolicy:selectedPolicy];

}

@end

File: pippip-native-ios/More/ContactPolicyCell.swift
//
//  ContactPolicyCell.swift
//  pippip-native-ios
//
//  Created by Steve Brenneis on 4/10/18.
//  Copyright © 2018 seComm. All rights reserved.
//

import UIKit

class ContactPolicyCell: UITableViewCell {

    @IBOutlet weak var contactPolicySwitch: UISwitch!

    var contactManager = ContactManager()
    var config = Configurator()
    var currentPolicy = "whitelist"
    var selectedPolicy = "whitelist"

    override func awakeFromNib() {
        super.awakeFromNib()
        // Initialization code
        currentPolicy = config.contactPolicy
        if (currentPolicy == "public") {
            contactPolicySwitch.setOn(true, animated: true)
        }
        else {
            contactPolicySwitch.setOn(false, animated: true)
        }

        NotificationCenter.default.addObserver(self, selector: #selector(policyUpdated(_:)),
                                               name: Notifications.PolicyUpdated, object: nil)

    }

    override func setSelected(_ selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)

        // Configure the view for the selected state
    }

    @objc class func cellItem() -> MoreCellItem {
        
        let item = MoreCellItem()
        item.cellHeight = 65.0
        item.cellReuseId = "ContactPolicyCell"
        return item
        
    }

    @objc func policyUpdated(_ notification: Notification) {

        guard let info = notification.userInfo else { return }
        if let result = info["result"] as? String {
            DispatchQueue.main.async {
                if (result == "policySet") {
                    self.currentPolicy = self.selectedPolicy
                    self.config.contactPolicy = self.currentPolicy
                    var info = [AnyHashable: Any]()
                    info["policy"] = self.currentPolicy
                    NotificationCenter.default.post(name: Notifications.PolicyChanged, object: nil,
                                                    userInfo: info)
                }
                self.resetCell()
            }
        }
        else {
            var info = [AnyHashable: Any]()
            info["title"] = "Policy Error"
            info["message"] = "Invalid server response"
            NotificationCenter.default.post(name: Notifications.PresentAlert, object: nil, userInfo: info)
            DispatchQueue.main.async {
                self.resetCell()
            }
        }

    }

    func resetCell() {

        if (currentPolicy == "public") {
            contactPolicySwitch.setOn(true, animated: true)
        }
        else {
            contactPolicySwitch.setOn(false, animated: true)
        }

    }

    @IBAction func policyChanged(_ sender: UISwitch) {

        if (sender.isOn) {
            selectedPolicy = "public"
        }
        else {
            selectedPolicy = "whitelist"
        }
        contactManager.setContactPolicy(selectedPolicy)

    }

}
